wallet/src/repository/wallet/mod.rs

```diff
@@ -950,153 +1020,187 @@ where
+            utxo_strategy,
+        )?;
+
+        Ok((inputs, outputs))
     }
 
     fn create_dr_transaction_components(
         &self,
         state: &mut State,
-        value: u64,
+        request: DataRequestOutput,
         fee: u64,
-    ) -> Result<types::TransactionComponents> {
-        self.create_transaction_components(state, value, fee, None, true)
+    ) -> Result<(Vec<Input>, Vec<ValueTransferOutput>)> {
+        let utxo_strategy = UtxoSelectionStrategy::Random;
+        let tx_pending_timeout = u64::from(state.epoch_constants.checkpoints_period) * 10;
+        let timestamp = u64::try_from(get_timestamp()).unwrap();
+
+        let (inputs, outputs) = self.build_inputs_outputs_inner_wallet(
+            vec![],
+            Some(&request),
+            fee,
+            state,
+            timestamp,
+            tx_pending_timeout,
+            None,
+            utxo_strategy,
+        )?;
+
+        Ok((inputs, outputs))
     }
 
-    fn create_transaction_components(
+    #[allow(clippy::too_many_arguments)]
+    fn build_inputs_outputs_inner_wallet(
         &self,
-        state: &mut State,
-        value: u64,
+        outputs: Vec<ValueTransferOutput>,
+        dr_output: Option<&DataRequestOutput>,
         fee: u64,
-        recipient: Option<(types::PublicKeyHash, u64)>,
-        // When creating data request transactions, the change address must be the same as the
-        // first input address
-        change_address_same_as_input: bool,
-    ) -> Result<types::TransactionComponents> {
-        let target = value.saturating_add(fee);
-        let mut payment = 0u64;
-        let mut inputs = Vec::with_capacity(5);
-        let mut outputs = Vec::with_capacity(2);
-        let mut sign_keys = Vec::with_capacity(5);
-        let mut used_utxos = Vec::with_capacity(5);
-        let mut balance = state.balance;
+        state: &mut State,
+        timestamp: u64,
+        tx_pending_timeout: u64,
+        // The block number must be lower than this limit
+        block_number_limit: Option<u32>,
+        utxo_strategy: UtxoSelectionStrategy,
+    ) -> Result<(Vec<Input>, Vec<ValueTransferOutput>)> {
+        let mut wallet_utxos = WalletUtxos { state };
+
+        // On error just assume the value is u64::max_value(), hoping that it is
+        // impossible to pay for this transaction
+        let output_value: u64 = transaction_outputs_sum(&outputs)
+            .unwrap_or(u64::max_value())
+            .saturating_add(
+                dr_output
+                    .map(|o| o.checked_total_value().unwrap_or(u64::max_value()))
+                    .unwrap_or_default(),
+            );
 
-        if let Some((pkh, time_lock)) = recipient {
-            outputs.push(types::VttOutput {
-                pkh,
-                value,
-                time_lock,
-            });
-        }
+        let (output_pointers, input_value) = take_enough_utxos(
+            &mut wallet_utxos,
+            output_value + fee,
```

Comment:
This also needs overflow checks, and `output_value` should use `checked_add`, same as in the node